networkx/algorithms/approximation/tests/test_steinertree.py

```diff
@@ -190,6 +190,17 @@ def test_multigraph_steiner_tree(self):
             S = steiner_tree(G, terminal_nodes, method=method)
             assert edges_equal(S.edges(data=True, keys=True), expected_edges)
 
+    def test_remove_nonterminal_leaves(self):
+        from networkx.algorithms.approximation.steinertree import (
+            _remove_nonterminal_leaves,
+        )
+
+        G = nx.Graph()
+        G.add_edges_from([(1, 2), (2, 3), (3, 4), (4, 5)], weight=1)
+        _remove_nonterminal_leaves(G, [2, 3])
+
+        assert list(G.nodes) == [2, 3]  # only the terminal nodes are left
```

Comment:
One minor suggestion to beef up the test so that there are *multiple* non-terminal nodes on both ends of the path:

```suggestion
        G = nx.path_graph(10)
        nx.approximation.steinertree._remove_nonterminal_leaves(G, [4, 5, 6])

        assert list(G) == [4, 5, 6]  # only the terminal nodes are left
```

Reply:
OOh yeah I like that, I just wrote it to display as simple as possible the deficiencies in the current code. This more profoundly demonstrates it though! I'm onboard!